src/index.js

```diff
@@ -1,98 +1,140 @@
+ * @param {number} [options.errorPropagation = 50] - How many evaluations (per point per step) of the fitted function to use to approximate the error propagation through it
+ * @return {Result}
  */
-export default function levenbergMarquardt(
-  data,
-  parameterizedFunction,
-  options = {}
-) {
+export default function levenbergMarquardt(data, paramFunction, options = {}) {
   let {
     maxIterations = 100,
-    gradientDifference = 10e-2,
-    damping = 0,
-    errorTolerance = 10e-3,
-    minValues,
+    gradientDifference = 1e-6,
+    damping = 0.1,
+    dampingDrop = 0.1,
+    dampingBoost = 1.5,
+    maxDamping = Number.MAX_SAFE_INTEGER,
+    minDamping = Number.EPSILON,
     maxValues,
-    initialValues
-  } = options;
+    minValues,
+    residualEpsilon = 1e-6,
+    initialValues,
+    errorPropagation = 50
+  } = legacy.compatOptions(options);
 
   if (damping <= 0) {
     throw new Error('The damping option must be a positive number');
-  } else if (!data.x || !data.y) {
-    throw new Error('The data parameter must have x and y elements');
+  } else if (!data || !data.x || !data.y) {
+    throw new Error('The data object must have x and y elements');
   } else if (
     !Array.isArray(data.x) ||
     data.x.length < 2 ||
     !Array.isArray(data.y) ||
     data.y.length < 2
   ) {
+    throw new Error('The data must have more than 2 points');
+  } else if (data.x.length !== data.y.length) {
     throw new Error(
-      'The data parameter elements must be an array with more than 2 points'
+      'The data object must have equal number of x and y coordinates'
     );
-  } else if (data.x.length !== data.y.length) {
-    throw new Error('The data parameter elements must have the same size');
   }
 
-  var parameters =
-    initialValues || new Array(parameterizedFunction.length).fill(1);
-  let parLen = parameters.length;
+  let params = initialValues || new Array(paramFunction.length).fill(1);
+  let parLen = params.length;
   maxValues = maxValues || new Array(parLen).fill(Number.MAX_SAFE_INTEGER);
   minValues = minValues || new Array(parLen).fill(Number.MIN_SAFE_INTEGER);
 
-  if (maxValues.length !== minValues.length) {
-    throw new Error('minValues and maxValues should have the same size');
+  if (!Array.isArray(params)) {
+    throw new Error('initialValues must be an array');
   }
 
-  if (!Array.isArray(parameters)) {
-    throw new Error('initialValues must be an array');
+  if (
+    maxValues.length !== minValues.length ||
+    maxValues.length !== params.length
+  ) {
+    throw new Error(
+      'minValues and maxValues must have the same size as the number of parameters'
+    );
   }
 
-  var error = errorCalculation(data, parameters, parameterizedFunction);
+  initializeErrorPropagation(errorPropagation);
+
+  /** @type Array<number> */
+  var residualDifferences = Array(10).fill(NaN);
 
-  var converged = error <= errorTolerance;
+  var residuals = sumOfSquaredResiduals(data, params, paramFunction);
```

Comment:
We seem to be mixing `var` and `let`/`const`. I don't know if the switch belongs in this PR, but I think we should migrate away from using `var` unless we're trying to be compatible with really old engines. (And with the move to TS, that wouldn't be a problem anyway since it could just be translated down at the end.)

Reply:
We can migrate to let and const with typescript

Reply:
I'm used to `let` and `const` from TS, so it's possible that I accidentaly used them instead of `var` on some places.